pkg/network/http/incomplete_stats.go

```diff
@@ -0,0 +1,123 @@
+// (eg. httpTX objects that have either only the request or response information)
+// this happens only in the context of localhost traffic with NAT.
+// Imagine, for example, that you have two containers in the same host:
+// * Container A: 1.1.1.1
+// * Container B: 3.3.3.3
+// And a NAT rule: 2.2.2.2 -> 3.3.3.3
+// Now let's say Container A issues a HTTP request to 2.2.2.2;
+// The eBPF socket filter program will see two "disjoint" TCP segments:
+// (1.1.1.1:ephemeral-port -> 2.2.2.2:server-port) GET / HTTP/1.1
+// (3.3.3.3 -> 1.1.1.1:ephemeral-port) HTTP/1.1 200 OK
+// Because of that, we join these two parts of the transaction here in userspace using the
+// client address (1.1.1.1:ephemeral-port)
+// There is, however, another variable that can further complicate this: keep-alives
+// You could have, for example, one TCP socket issuing multiple requests:
+// t0: (1.1.1.1:ephemeral-port -> 2.2.2.2:server-port) GET / HTTP/1.1
+// t1: (3.3.3.3 -> 1.1.1.1:ephemeral-port) HTTP/1.1 200 OK
+// t2: (1.1.1.1:ephemeral-port -> 2.2.2.2:server-port) GET / HTTP/1.1
+// t3: (3.3.3.3 -> 1.1.1.1:ephemeral-port) HTTP/1.1 200 OK
+// The problem is, due to the way our eBPF batching works, there is no guarantee that these
+// incomplete events will be read in the order they happened, so if we had a greedy approach
+// that joined events as soon as they're sent from eBPF, we could potentially join
+// request segment at "t0" with response segment "t3". This is why we buffer data here for 30 seconds
+// and then sort all events by their timestamps before joining them.
+type incompleteBuffer struct {
+	data       map[Key]*txParts
+	maxEntries int
+	telemetry  *telemetry
+}
+
+type txParts struct {
+	requests  []httpTX
+	responses []httpTX
+}
+
+func newIncompleteBuffer(c *config.Config, telemetry *telemetry) *incompleteBuffer {
+	return &incompleteBuffer{
+		data:       make(map[Key]*txParts),
+		maxEntries: c.MaxHTTPStatsBuffered,
+		telemetry:  telemetry,
+	}
+}
+
+func (b *incompleteBuffer) Add(tx httpTX) {
+	key := Key{
+		SrcIPHigh: uint64(tx.tup.saddr_h),
+		SrcIPLow:  uint64(tx.tup.saddr_l),
+		SrcPort:   uint16(tx.tup.sport),
+	}
+
+	parts, ok := b.data[key]
+	if !ok {
+		if len(b.data) >= b.maxEntries {
+			atomic.AddInt64(&b.telemetry.dropped, 1)
+			return
+		}
+
+		parts = &txParts{
+			requests:  make([]httpTX, 0, 5),
+			responses: make([]httpTX, 0, 5),
+		}
+
+		b.data[key] = parts
+	}
+
+	if tx.StatusClass() == 0 {
+		parts.requests = append(parts.requests, tx)
+	} else {
+		parts.responses = append(parts.responses, tx)
+	}
+}
+
+func (b *incompleteBuffer) Flush() []*httpTX {
+	var joined []*httpTX
+
+	for _, parts := range b.data {
+		// TODO: in this loop we're sorting all transactions at once, but we could also
+		// consider sorting data during insertion time (using a tree-like structure, for example)
+		sort.Sort(byRequestTime(parts.requests))
+		sort.Sort(byResponseTime(parts.responses))
```

Comment:
This appears to match simply based on request start time and the last response time. Couldn't that result in mismatches if you have one requested nested within another. Example:
A req
B req
B resp
A resp

This would try to match A req to B resp and B req to A resp.

Reply:
I don't think that's is possible for the same source port 🤔

Reply:
If the source port is the server port, isn't that possible?

Reply:
sorry, in the context of HTTP the source port is always the Client port (we normalize the tuples on kernel space) (we should probably use client/server nomenclature to avoid confusion 😬 )

Reply:
I see we do that based on a port not being in the ephemeral range. If both ports are in the ephemeral range, should we do something like normalize to have the higher port be the client? Otherwise, they will not match correct?

Reply:
sure, that sounds a contrived scenario for a HTTP server, but I guess that's within the realms of possibility.
I would just argue that this normalization should be done in the kernel side, because right now if both ports are in the ephemeral range HTTP monitoring may not work as it is, because there is a chance you won't be able to tie the request and response even if there is no localhost, NAT or keep-alives in the mix.

Reply:
> that sounds a contrived scenario for a HTTP server,

I think it is more likely in a containerized environment with bin-packing and service discovery. Using well known hard-coded ports is less-likely.

> normalization should be done in the kernel side

Of course.

Reply:
Added this change on a separate PR https://github.com/DataDog/datadog-agent/pull/10457